File: integrationtest/self-healing-integration-tests/src/main/java/de/dkeiss/taxishare/pages/LoginPage.java
package de.dkeiss.taxishare.pages;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;

public class LoginPage extends AbstractPage {

    public static final String URL = "login";
    public static final String URL_2 = "loginV2";
    public static final String URL_3 = "loginV3";
    public static final String URL_4 = "loginV4";

    public LoginPage(WebDriver driver) {
        super(driver);
    }

    public String getURL() {
        return URL;
    }

    @FindBy(name = "username")
    private WebElement usernameInput;

    @FindBy(name = "password")
    private WebElement passwordInput;

    @FindBy(className = "btn-primary")
    private WebElement submitButton;

    public void login(String username, String password) {
        usernameInput.sendKeys(username);
        passwordInput.sendKeys(password);
        submitButton.click();
    }

}

File: integrationtest_python/self_healing_integration_tests/features/pages/login_page.py
from __future__ import annotations

from dataclasses import dataclass

from selenium.webdriver.common.by import By

from .base_page import BasePage
from ..support.self_healing import HealableLocator


@dataclass
class LoginPage(BasePage):
    url_path: str = "login"

    username_input: HealableLocator = HealableLocator(By.NAME, "username", "Username input field")
    password_input: HealableLocator = HealableLocator(By.NAME, "password", "Password input field")
    submit_button: HealableLocator = HealableLocator(By.CSS_SELECTOR, "button.btn-primary", "Login button")

    def login(self, username: str, password: str) -> None:
        username_element = self.find(self.username_input)
        username_element.clear()
        username_element.send_keys(username)

        password_element = self.find(self.password_input)
        password_element.clear()
        password_element.send_keys(password)

        self.find(self.submit_button).click()

